### PR TITLE
MoE: correct + faster a16w4 (bf16 A x MXFP4 W) SiTUv2 kernel (replaces old a16w4)

### PR DESCRIPTION
**Pull request by coderfeli**
## Summary

**Replaces** aiter's a16w4 (bf16 activation x MXFP4 weight) SiTUv2 fused-MoE
kernel with a numerically-correct, faster ported FlyDSL 2-stage kernel
(`aiter/ops/flydsl/kernels/moe_2stage_a16wmix`), and **removes** the old
`compile_mixed_moe_gemm{1,2}_a16w4` path entirely.

aiter's previous a16w4 SiTUv2 kernel **fails aiter's own strict accuracy gate**
(`op_tests/test_moe_2stage.py`, bf16 SiTUv2 reference): `logits_diff ~1.0`
(essentially wrong output). The port **passes the same gate at `logits_diff
~1.5e-5`** across the full token sweep on both tuned Kimi shapes (3584x384,
3584x512, E896, topk16).

## Reproduction (independent re-measurement, gfx950)

Re-measured the shipped port against **its own tuned CSV**
(`kimik3_fp4_tuned_fmoe.csv`, `abf16_wfp4`) on gfx950 (cu=256) via
`test_moe_2stage.py --kernel --no-legacy --csv-filter abf16_wfp4`, both Kimi
shapes. The kernel runs correctly and **meets or beats its tuned latency at
every token** (0.80–1.06×), monotonic across M. Ratio = measured / tuned-CSV
(lower = faster than the recorded tune).

Correctness: `op_tests/flydsl_tests/test_flydsl_moe_a16wfp4.py` **6/6 pass**
(2 shapes × tok 1/16/128) vs the authoritative torch SiTUv2 reference.

### 3584x512
| tok | measured total (µs) | tuned-CSV (µs) | ratio |
|--:|--:|--:|--:|
| 1 | 19.9 | 24.0 | 0.83× |
| 2 | 29.7 | 33.3 | 0.89× |
| 3 | 38.9 | 42.5 | 0.92× |
| 4 | 41.4 | 48.2 | 0.86× |
| 8 | 62.2 | 77.5 | 0.80× |
| 16 | 111.8 | 140.1 | 0.80× |
| 32 | 200.0 | 222.4 | 0.90× |
| 64 | 307.0 | 330.1 | 0.93× |
| 128 | 424.8 | 428.1 | 0.99× |
| 256 | 479.2 | 483.2 | 0.99× |
| 512 | 470.4 | 508.2 | 0.93× |
| 1024 | 524.5 | 541.6 | 0.97× |
| 2048 | 843.9 | 792.8 | 1.06× |
| 4096 | 1232.2 | 1215.6 | 1.01× |
| 8192 | 2021.1 | 2025.7 | 1.00× |
| 16384 | 3669.9 | 3694.2 | 0.99× |

### 3584x384
| tok | measured total (µs) | tuned-CSV (µs) | ratio |
|--:|--:|--:|--:|
| 1 | 19.0 | 21.6 | 0.88× |
| 2 | 25.8 | 28.5 | 0.91× |
| 3 | 34.0 | 38.9 | 0.87× |
| 4 | 36.9 | 42.7 | 0.86× |
| 8 | 56.8 | 59.8 | 0.95× |
| 16 | 91.5 | 107.3 | 0.85× |
| 32 | 162.6 | 177.0 | 0.92× |
| 64 | 230.9 | 260.6 | 0.89× |
| 128 | 314.0 | 336.0 | 0.93× |
| 256 | 354.1 | 374.7 | 0.95× |
| 512 | 354.7 | 392.7 | 0.90× |
| 1024 | 391.3 | 422.5 | 0.93× |
| 2048 | 644.4 | 651.3 | 0.99× |
| 4096 | 994.1 | 1087.3 | 0.91× |
| 8192 | 1689.6 | 1994.4 | 0.85× |
| 16384 | 3081.7 | 3600.3 | 0.86× |

## What changed

- **Vendored kernel package** `aiter/ops/flydsl/kernels/moe_2stage_a16wmix/`
  (3 files: `gemm1.py`, `gemm2.py`, `__init__.py`): gemm1 (gate+up GEMM + SiTUv2
  -> sorted bf16 intermediate), gemm2 (down-proj + routing-weighted atomic bf16
  scatter). Non-scaled `MFMA(16,16,32,bf16)`, CDNA (gfx942/gfx950). Consolidated
  layout (shared helpers in gemm1, launch glue + CSV tile resolvers in
  `__init__`); scale loads use the `make_buffer_tensor`/`BufferCopy32b` layout API.
- **`fused_moe_` dispatch** (`aiter/ops/flydsl/moe_2stage_a16w4_dispatch.py`):
  serves bf16-a16w4 SiTUv2 (the kernel is gate-mode agnostic). Converts the
  incoming guinterleave (a8w4/mxfp8-shared) weight/scale layout to the port's
  standard `shuffle_weight`/`e8m0_shuffle` layout via pure byte permutations
  (verified round-trip). gemm2 tiles come from aiter's tuned CSV
  (`resolve_a16w4_gemm2_config`, on-par-to-faster); gemm1 keeps its own tuned
  default — a perf-gate showed the CSV gemm1 tiles regress mid/high-M by
  1.1x–2.0x, so they are not taken.
- **Removed the old a16w4 path**: `compile_mixed_moe_gemm{1,2}_a16w4`, the compat
  wrapper `compile_a16w4_moe_gemm2`, and the a16w4-only `_decode_e8m0_byte_to_f32`
  helper (~4076 lines) from `mixed_moe_gemm_2stage.py`; the a16w4 branches of
  `compile_flydsl_moe_stage1/2`; and the dead a16w4 sub-clause / `_a_type`
  bf16-fp16 branches in `get_2stage_cfgs`. Generic `compile_mixed_moe_gemm1/2`
  and all other dtypes (a8w4 / int4 / mxfp8 / a4w4) are untouched.
- **Fallback handling**: features the port does not implement (non-default SiTUv2
  beta/linear_beta, per-expert bias, expert-parallel masking) raise a clear
  `NotImplementedError` rather than routing to the numerically-broken old kernel.
- **Tests**: aiter's original `op_tests/test_moe_2stage.py` a16w4 (`abf16_wfp4`)
  SiTUv2 rows now exercise the new kernel through the production dispatch —
  `_effective_gate_mode` returns SEPARATED for bf16-a16w4 (previously INTERLEAVE,
  which silently downgraded those rows to a8w4/fp8), and the a16w4 SiTUv2 path is
  bias-free (the port has no per-expert bias). No bespoke test file.

### CODE REVIEW

File: aiter/ops/flydsl/moe_kernels.py [hunk 1]
```diff
@@ -1460,12 +1453,53 @@ def _flydsl_moe_stage1_impl(
...
+    _is_a16w4 = a_dtype == "bf16" and b_dtype == "fp4"
+    if _is_a16w4:
+        # a16w4 (bf16 A x fp4 W): ported gemm1 -> bf16 sorted intermediate,
+        # threaded to stage2 unchanged. Tiles from the CSV kernelName;
+        # waves_per_eu=None (a no-_w name parses to wpe=1, a different kernel).
+        from aiter.ops.flydsl.kernels.moe_2stage_a16wmix import flydsl_a16w4_gemm1
+
+        _act = "situv2" if act in ("situv2", "situ") else act
+        sorted_size = int(sorted_expert_ids.shape[0]) * int(tile_m)
+        _alloc = torch.zeros if inter_dim_pad > 0 else torch.empty
```
Comment (valarLip) [inline aiter/ops/flydsl/moe_kernels.py:1465]: zeros ?

File: aiter/ops/flydsl/moe_kernels.py [hunk 2]
```diff
@@ -1460,12 +1453,53 @@ def _flydsl_moe_stage1_impl(
...
+        inter_sorted = _alloc(sorted_size, inter_dim, dtype=torch.bfloat16, device=dev)
+        flydsl_a16w4_gemm1(
+            a_bf16=a.to(torch.bfloat16).contiguous(),
```
Comment (valarLip) [inline aiter/ops/flydsl/moe_kernels.py:1468]: contiguous?

Comment (coderfeli) [reply aiter/ops/flydsl/moe_kernels.py:1468]: no real call. just make sure no error. can be deleted.

File: aiter/ops/flydsl/moe_kernels.py (hunk 1 above)
Comment (coderfeli) [reply aiter/ops/flydsl/moe_kernels.py:1465]: ported from old code, https://github.com/ROCm/aiter/pull/4502/changes#diff-d5d52b4d0fede962b29252c6940ba9aa70479833fc6b767058574c607e140004L1496. seems for bug fixes of gptoss. No sure if needed for other models. will not be trigger for inter_dim_pad=0
